fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -413,22 +404,19 @@ def reduce(*_: Any) -> None:
                     self._grad_to_be_reduced[index] = False
                     param.grad.mul_(self.world_size_scaling)
 
-                    if self.reduce_fp16:
-                        param.grad.data = param.grad.data.half()
-
                     # Future work includes clearing up the buffer if possible
                     def cleanup() -> None:
                         if dst_rank != self.global_rank:
                             param.grad = None
-                        else:
-                            assert param.grad is not None
-                            param.grad.data = param.grad.data.to(dtype=param.dtype)
+
+                    if self.backend == dist.Backend.NCCL:
+                        torch.cuda.synchronize()
```

Comment:
doing that completely tanks the perf, but the results are not correct if I remove just this line (same for the 2 other same lines below).. cc @ngimel @mrshenli, any other way to make sure that the NCCL stream(s) wait for a given event on the current one ?